extensions/amp-animation/0.1/css-expr-ast.js

```diff
@@ -0,0 +1,250 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {endsWith} from '../../../src/string';
+
+
+/**
+ * A base class for all CSS node components defined in the
+ * `css-expr-impl.jison`.
+ * @abstract
+ */
+export class CssNode {
+  constructor() {}
+}
+
+
+/**
+ * A concatenation of CSS expressions: `translateX(...) rotate(...)`,
+ * `1s normal`, etc.
+ */
+export class CssConcatNode extends CssNode {
+  constructor() {
+    super();
+    /** @const @private {!Array<!CssNode>} */
+    this.array_ = [];
+  }
+
+  /**
+   * Concatenates two sets of expressions.
+   * @param {!CssNode} nodeOrSet
+   * @param {!CssNode} otherNodeOrSet
+   * @return {!CssConcatNode}
+   */
+  static concat(nodeOrSet, otherNodeOrSet) {
+    let set;
+    if (nodeOrSet instanceof CssConcatNode) {
+      set = nodeOrSet;
+    } else {
+      set = new CssConcatNode();
+      set.array_.push(nodeOrSet);
+    }
+    if (otherNodeOrSet instanceof CssConcatNode) {
+      set.array_.concat(otherNodeOrSet.array_);
```

Comment:
I'd like to avoid construction and state leaking into public methods of AST nodes. I'd rather it all be fully compressed to this module only and preserve full obfuscation of these details on compiler level.